File: MusicPlayer/Utilities/Youtube.cs
using System.Collections.Generic;
using System.Text.Json.Serialization;

namespace MusicPlayer
{
	public class YouTubeSearchResponse
	{
		[JsonPropertyName("kind")]
		public string? Kind { get; set; }

		[JsonPropertyName("etag")]
		public string? Etag { get; set; }

		[JsonPropertyName("nextPageToken")]
		public string? NextPageToken { get; set; }

		[JsonPropertyName("regionCode")]
		public string? RegionCode { get; set; }

		[JsonPropertyName("pageInfo")]
		public PageInfo? PageInfo { get; set; }

		[JsonPropertyName("items")]
		public List<SearchResultItem>? Items { get; set; }
	}

	public class PageInfo
	{
		[JsonPropertyName("totalResults")]
		public int TotalResults { get; set; }

		[JsonPropertyName("resultsPerPage")]
		public int ResultsPerPage { get; set; }
	}

	public class SearchResultItem
	{
		[JsonPropertyName("kind")]
		public string? Kind { get; set; }

		[JsonPropertyName("etag")]
		public string? Etag { get; set; }

		[JsonPropertyName("id")]
		public required VideoId Id { get; set; }

		[JsonPropertyName("snippet")]
		public VideoSnippet? Snippet { get; set; }
	}

	public class VideoId
	{
		[JsonPropertyName("kind")]
		public required string Kind { get; set; }

		[JsonPropertyName("videoId")]
		public required string Id { get; set; }
	}

	public class VideoSnippet
	{
		[JsonPropertyName("publishedAt")]
		public string? PublishedAt { get; set; }

		[JsonPropertyName("channelId")]
		public string? ChannelId { get; set; }

		[JsonPropertyName("title")]
		public string? Title { get; set; }

		[JsonPropertyName("description")]
		public string? Description { get; set; }

		[JsonPropertyName("thumbnails")]
		public Thumbnails? Thumbnails { get; set; }

		[JsonPropertyName("channelTitle")]
		public string? ChannelTitle { get; set; }

		[JsonPropertyName("liveBroadcastContent")]
		public string? LiveBroadcastContent { get; set; }

		[JsonPropertyName("publishTime")]
		public string? PublishTime { get; set; }
	}

	public class Thumbnails
	{
		[JsonPropertyName("default")]
		public Thumbnail? Default { get; set; }

		[JsonPropertyName("medium")]
		public Thumbnail? Medium { get; set; }

		[JsonPropertyName("high")]
		public Thumbnail? High { get; set; }
	}

	public class Thumbnail
	{
		[JsonPropertyName("url")]
		public string? Url { get; set; }

		[JsonPropertyName("width")]
		public int Width { get; set; }

		[JsonPropertyName("height")]
		public int Height { get; set; }
	}
}

File: app/src/main/java/com/gami13/musicplayer/Youtube.kt
package com.gami13.musicplayer

import android.content.Context
import androidx.datastore.core.DataStore
import androidx.datastore.preferences.core.Preferences
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.stringPreferencesKey
import androidx.datastore.preferences.preferencesDataStore
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import kotlinx.serialization.Serializable

val Context.dataStore: DataStore<Preferences> by preferencesDataStore(name = Constants.SETTINGS_NAME)
class SettingsRepository(private val context: Context) {

  companion object {
    private val MUSIC_DIRECTORY_KEY = stringPreferencesKey("music_directory")
  }

  val musicDirectory: Flow<String> = context.dataStore.data
    .map { preferences -> preferences[MUSIC_DIRECTORY_KEY]  ?: "" }


  suspend fun saveMusicDirectory(directoryUri: String) {
    context.dataStore.edit { preferences ->
      preferences[MUSIC_DIRECTORY_KEY] = directoryUri
    }
  }
}



@Serializable
data class YoutubeSearchResponse(
  val kind: String,
  val etag: String,
  val nextPageToken: String,
  val regionCode: String,
  val pageInfo: PageInfo,
  val items: List<SearchResultItem>
)

@Serializable
data class PageInfo(
  val totalResults: Int,
  val resultsPerPage: Int
)

@Serializable
data class SearchResultItem(
  val kind: String,
  val etag: String,
  val id: VideoId,
  val snippet: VideoSnippet
)

@Serializable
data class VideoId(
  val kind: String,
  val videoId: String
)

@Serializable
data class VideoSnippet(
  val publishedAt: String,
  val channelId: String,
  val title: String,
  val description: String,
  val thumbnails: Thumbnails,
  val channelTitle: String,
  val liveBroadcastContent: String,
  val publishTime: String
)

@Serializable
data class Thumbnails(
  val default: Thumbnail,
  val medium: Thumbnail,
  val high: Thumbnail
)

@Serializable
data class Thumbnail(
  val url: String,
  val width: Int,
  val height: Int
)

